basic/escrow/src/lib.rs

```diff
@@ -0,0 +1,120 @@
+use scrypto::prelude::*;
+
+blueprint! {
+    struct Escrow {
+        token_a: Vault,
+        token_b: Vault,
+        account_a_badge: ResourceDef,
+        account_b_badge: ResourceDef,
+        account_a_accepted: bool,
+        account_b_accepted: bool,
+        trade_canceled: bool
+    }
+
+    impl Escrow {
+        pub fn new(token_a_address: Address, token_b_address: Address) -> (Component, Bucket, Bucket) {
+            // Create the badges that will allow the component to authenticate the users
+            let account_a_badge = ResourceBuilder::new()
+                .metadata("symbol", "BADGE A")
+                .new_badge_fixed(1);
+
+            let account_b_badge = ResourceBuilder::new()
+                .metadata("symbol", "BADGE B")
+                .new_badge_fixed(1);
+
+            let account_a_badge_resource = account_a_badge.resource_def();
+            let account_b_badge_resource = account_b_badge.resource_def();
+
+            let component = Self {
+                token_a: Vault::new(token_a_address),
+                token_b: Vault::new(token_b_address),
+                account_a_badge: account_a_badge_resource,
+                account_b_badge: account_b_badge_resource,
+                account_a_accepted: false,
+                account_b_accepted: false,
+                trade_canceled: false
+            }
+            .instantiate();
+
+            (component, account_a_badge, account_b_badge)
+        }
+
+        // Allow the users to put the tokens they want to trade inside the component's vault
+        #[auth(account_a_badge, account_b_badge)]
+        pub fn put_tokens(&mut self, tokens: Bucket) {
+            let user_id = Self::get_user_id(&auth);
+
+            scrypto_assert!(!(self.account_a_accepted || self.account_b_accepted) , "Can't add more tokens when someone accepted");
+            scrypto_assert!(!self.trade_canceled, "The trade was canceled");
+
+            if user_id == self.account_a_badge.address() {
+                self.token_a.put(tokens);
+            } else {
+                self.token_b.put(tokens);
+            }
+        }
+
+        #[auth(account_a_badge)]
+        pub fn withdraw_all(&mut self) -> Bucket {
+            self.token_a.take_all()
+        }
```

Comment:
I think there is a bug here. Consider the following even sequence:
1. User A deposits 500 DRG;
1. User B deposits 10 XRD;
1. User B accepts;
1. User A accepts;
1. User A withdraw all 500 DRG;
1. User A withdraw 10 XRD;
1. There is nothing left for B to withdraw.

Better to remove it.

Reply:
I forgot to remove "withdraw_all". It wasn't used anymore.